xds/src/test/java/io/grpc/xds/XdsClientImplTest.java

```diff
@@ -0,0 +1,1373 @@
+                                "cluster-baz.googleapis.com"))))
+                .build()))));
+    DiscoveryResponse response =
+        buildDiscoveryResponse("0", listeners, XdsClientImpl.ADS_TYPE_URL_LDS, "0000");
+    responseObserver.onNext(response);
+
+    // Client sends an ACK LDS request.
+    verify(requestObserver)
+        .onNext(eq(buildDiscoveryRequest("0", "foo.googleapis.com:8080",
+            XdsClientImpl.ADS_TYPE_URL_LDS, "0000")));
+
+    ArgumentCaptor<Status> errorStatusCaptor = ArgumentCaptor.forClass(null);
+    verify(configWatcher).onError(errorStatusCaptor.capture());
+    Status error = errorStatusCaptor.getValue();
+    assertThat(error.getCode()).isEqualTo(Code.NOT_FOUND);
+    assertThat(error.getDescription())
+        .isEqualTo("Listener for requested resource [foo.googleapis.com:8080] does not exist");
+
+    verifyNoMoreInteractions(requestObserver);
+  }
+
+  /**
+   * An LDS response contains the requested listener and an in-lined RouteConfiguration message for
+   * that listener. But the RouteConfiguration message is invalid as it does not contain any
+   * VirtualHost with domains matching the requested hostname.
+   * The LDS response is NACKed, as if the XdsClient has not received this response.
+   * The config watcher is NOT notified with an error.
+   */
+  @Test
+  public void failToFindVirtualHostInLdsResponseInLineRouteConfig() {
+    xdsClient.watchConfigData(HOSTNAME, PORT, configWatcher);
+    StreamObserver<DiscoveryResponse> responseObserver = responseObservers.poll();
+    StreamObserver<DiscoveryRequest> requestObserver = requestObservers.poll();
+
+    // Client sends an LDS request for the host name (with port) to management server.
+    verify(requestObserver)
+        .onNext(eq(buildDiscoveryRequest("", "foo.googleapis.com:8080",
+            XdsClientImpl.ADS_TYPE_URL_LDS, "")));
+
+    RouteConfiguration routeConfig =
+        buildRouteConfiguration(
+            "route.googleapis.com",
+            ImmutableList.of(
+                buildVirtualHost(ImmutableList.of("something does not match"),
+                    "some cluster"),
+                buildVirtualHost(ImmutableList.of("something else does not match"),
+                    "some other cluster")));
+
+    List<Any> listeners = ImmutableList.of(
+        Any.pack(buildListener("foo.googleapis.com:8080", /* matching resource */
+            Any.pack(HttpConnectionManager.newBuilder().setRouteConfig(routeConfig).build()))));
+    DiscoveryResponse response =
+        buildDiscoveryResponse("0", listeners, XdsClientImpl.ADS_TYPE_URL_LDS, "0000");
+    responseObserver.onNext(response);
+
+    // Client sends an NACK LDS request.
+    verify(requestObserver)
+        .onNext(
+            argThat(new DiscoveryRequestMatcher("", "foo.googleapis.com:8080",
+                XdsClientImpl.ADS_TYPE_URL_LDS, "0000")));
+
+    verifyZeroInteractions(configWatcher);
```

Comment:
For mockito 2.x `verifyZeroInteractions` is exactly the same as `verifyNoMoreInteractions`. Use `verify(configWatcher, never()).onConfigChanged` and `verify(configWatcher, never()).onError`  instead.

Reply:
What's the problem for `verifyZeroInteractions` or `verifyNoMoreInteractions`? If you never verified interactions of a mock, zero interactions indeed is no more interactions. `verifyZeroInteractions` is much more elegant than multiple lines of `never()` with `any()` matcher.

Reply:
Changed to verify with never.